### PR TITLE
[sphinx] Backport forgotten updates during the migration & other improvements

### LINKED ISSUE #7123: Find all regressions of Sphinx migration due to porting outdated chapters.
What remains to be done to complete the Sphinx migration?

For my part, I have reviewed all the post-beta migration PRs (using each time the second commit to try to spot regressions that could have happened during the migration). I haven't done so for what was merged before the beta.

There is a real risk of having lost things during the migration because the people who ported the chapters often started from an outdated version. Even if I have already spotted some of these regressions, I may have missed some (in particular small changes which are not a complete addition / removal of a paragraph). To make sure that there are no such regressions, the only way is IMHO for each ported chapter:
1. Determine what version was used for porting the chapter (8.7.0, 8.7.1, 8.7.2, more recent or even earlier?)
2. Take the commit log between this version (or the diff), e.g.: `git diff V8.7.0..1f2a922d52251f79a11d75c2205e6827a07e591b -- doc/refman/RefMan-xxx.tex`
3. Ensure (preferably on the rendered version) that the ported chapter did not lose any of these updates.

This is an important work which can be split between developers.

Other things to check before the final release: *moved to the GitHub project*

### PR DESCRIPTION
**Pull request by Zimmi48**
Backport forgotten updates during the migration + some other Sphinx fixes.

**Kind:** documentation

Fixes / closes #7123.

It was less bad than I expected.

### CODE REVIEW

**Comment by Zimmi48** [comment]
A more extensive check would be needed however for the chapters which were ported a very long time ago by @cpitclaudel. I noticed at least this problem: https://github.com/coq/coq/pull/7036#issuecomment-376924781 (still a TODO).

**Comment by Zimmi48** [comment]
Copying three remarks from the commit messages so as to get feedback:

- TODO: clarify why there is `@num`, `@natural` and `@integer`.

- Note that most error messages are not listed with their ending dot in the reference manual. So, by consistency, we should remove the few remaining ending dots. But for error messages that are made of several sentences, it does not look right to not have the ending dot. So should it be instead that they are all listed with their ending dot by consistency?

- TODO: clarify if we want to keep trailing dots in `cmd::` and make the practice consistent.

**Comment by cpitclaudel** [comment]
Thanks for this impressive work :heart_eyes: 

> TODO: clarify why there is @num, @natural and @integer.

I'd assume it's a historical artifact.

> So should it be instead that they are all listed with their ending dot by consistency?

I'd say yes :)

> clarify if we want to keep trailing dots in cmd:: and make the practice consistent.

I'd remove them.

**Comment by cpitclaudel** [comment]
MMaybe `@num`? Ideally we'd distinguish `@num`, `@nonegative`, `@positive`, but I don't know how accurate the current annotations are and whether they are worth keeping.

**Comment by Zimmi48** [comment]
About the chapters which were ported by you a long time ago, is the following list accurate?
- Syntax extensions
- Extended pattern-matching
- Universe Polymorphism

I know that @mattam82 backported changes for the universe polymorphism chapter. It seems that @maximedenes did it for the syntax extension one but there is no record for this. And the extended pattern-matching one was obviously forgotten (see my comment above) so there is still work to do wrt this chapter.

**Comment by cpitclaudel** [comment]
I converted these and 'The Gallina specification language' (plus some bits of other documents at  http://web.mit.edu/cpitcla/www/coq-rst/); I'm not sure which ones were used in the final release, though.

**Comment by Zimmi48** [comment]
OK so waiting for information from @maximedenes.

**Comment by Zimmi48** [comment]
So in fact, I checked again and my comment about a paragraph missing was wrong: it was between `\iffalse` and `\fi`.

**Comment by herbelin** [comment]
> So which one should we keep? @num only? I wonder whether @natural and @integer have the same semantics (non-negative integer) or not.

Here is a quick survey:

The parser distinguishes `natural` (unsigned machine integer), `integer` (signed machine integer) and `bigint` (mathematical _natural_ number) (see [`g_parsing.ml4`](https://github.com/coq/coq/blob/b1d749e59444f86e40f897c41739168bb1b1b9b3/parsing/g_prim.ml4#L112-L120)).

At the level of generic arguments, `integer` is bound to `int` (see [`stdarg.ml`](https://github.com/coq/coq/blob/b1d749e59444f86e40f897c41739168bb1b1b9b3/interp/stdarg.ml#L25-L26) and [`pcoq.ml`](https://github.com/coq/coq/blob/b1d749e59444f86e40f897c41739168bb1b1b9b3/parsing/pcoq.ml#L630), as used e.g. in the rules for `auto n`, see [`g_auto.ml4`](https://github.com/coq/coq/blob/b1d749e59444f86e40f897c41739168bb1b1b9b3/plugins/ltac/g_auto.ml4#L90)).

On its side, the reference manual defines `num` for natural and `int` for integer (in the gallina chapter).

Some tactics takes in practice an integer (e.g. `auto`), even though documented with a natural (with refman name `num`), and in practise behave strangely when called with a negative number.

So, clearly, for better documentation, whatever they are named exactly, we should uniformly have only `natural` and `integer` I guess (and, personally, I would be inclined to use only the mathematical ones). Also, for more robust behavior, a pass should be made on the use of numbers in the code...

**Comment by Zimmi48** [comment]
Thanks for this analysis @herbelin! Indeed a pass should be done to understand whether the uses of `integer` in parsing really make sense (in most case, I bet they don't).
With respect to the manual, there were only two instances of `integer`: in `Set Printing Width` and `Set Printing Depth`. I've just tested to see that with a negative number, these two commands are no-op's. So I'm in favor of pretending in the manual that all commands only take natural numbers (`num` since this is used the majority of the time already and the dual `integer` won't appear anymore).

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: Adding this index entry reveals a bug:
The next index entry `Set (cmd)` now appears as `(cmd)`.
On the other hand, if I put `Set (sort)` here, `Set (cmd)` is displayed normally but `Set (sort)` is displayed as `(sort)`.
I have observed the same bug with the two indices for `term` (one of them being added in this PR).

@cpitclaudel @maximedenes Can you look into this?

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: This does not appear in the symbol section of the index but in a new `_` section. Not great but not a critical problem either.

**Comment by Zimmi48** [comment]
@maximedenes I have removed the commit adding the index entries and revealing a bug to avoid delaying any longer the merge of this large PR.

**Comment by Zimmi48** [comment]
All the tests had passed but since now there was a conflict and I had to rebase, I pushed even more fixes.

**Comment by Zimmi48** [comment]
Note that there is still lots and lots to fix, just browsing the rst code is enough to find many problems. But this PR has already grown way too big so I can't continue adding to it. Moreover, the longer we wait until this is merged, the more conflicts there will be (in particular, because of the parts that were re-indented).

**Comment by Zimmi48** [comment]
Elpi failure is because it is currently broken with master.

**Comment by Zimmi48** [comment]
@maximedenes Could you merge this (and #7374)? It's only fixes, it shouldn't require an extensive review. I'm a bit sick of having had this PR opened for 20 days. Especially given the size of the PR. It prevents me working on any additional fixes to the manual.

**Comment by maximedenes** [comment]
Working on it now, sorry for the delay.

**Review by maximedenes — APPROVED** [review]
Ok, I had a quick scan at the whole patch. Of course, I can't review it precisely, but it looks good. Thanks!

**Comment by ejgallego** [comment]
Congrats folks on the huge and not too rewarding effort!

**Comment by ejgallego** [comment]
Well at least I understand we are in sync now, is that correct?

**Comment by Zimmi48** [comment]
Yes \o/

**Pull request merged**
